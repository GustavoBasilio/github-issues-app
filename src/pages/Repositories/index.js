import React, { useState, useEffect } from 'react';
import Icon from 'react-native-vector-icons/FontAwesome';
import { Keyboard } from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';

import api from '../../services/api';
import Repository from '../../components/Repository';
import { black } from '../../constants/colors';
import {
  Container,
  Form,
  Input,
  Submit,
  List,
} from './styles';

const Repositories = () => {

  const [insert, setInsert] = useState('');
  const [error, setError] = useState(false);
  const [repositories, setRepositories] = useState([]);
  const [loading, setLoading] = useState(true);


  /**
   * Load the repositories saved in the AsyncStorage
   *
   */
  async function loadRepositories() {
    setLoading(true);
    const result = JSON.parse(await AsyncStorage.getItem('repositories'));
    setRepositories(result);
    setLoading(false);
  }

  useEffect(() => {
    loadRepositories();
  }, []);


  /**
   * Save the repository in the the hook and AsyncStorage
   *
   * @param {Object} repository
   */
  async function saveRepository(repository) {

    const data = {
      id: repository.id,
      name: repository.name,
      owner: repository.owner.login,
      avatar: repository.owner.avatar_url,
    };

    const list = JSON.parse(await AsyncStorage.getItem('repositories'));

    // Check if the list saved is empty
    if (list === null) {
      await AsyncStorage.setItem('repositories', JSON.stringify([data]));
      setRepositories([data]);
    } else {

      const duplicated = list.filter((item) => item.id === data.id).length > 0;

      // Check if it is duplicated
      if (!duplicated) {
        await AsyncStorage.setItem('repositories', JSON.stringify([...list, data]));
        setRepositories([...list, data]);
      } else {
        setError(true);
      }

    }

    Keyboard.dismiss();

  }


  /**
   *  Fetch the repository data in the api then call the save function
   *
   */
  async function handleInsert() {

    try {

      setLoading(true);
      const result = await api.get(`/repos/${insert}`);
      await saveRepository(result.data);
      setError(false);
      setLoading(false);


    } catch (err) {

      setLoading(false);
      setError(true);
      console.log(err);

    }

  }

  return (
    <Container>
      <Form>
        <Input
          autoCapitalize="none"
          error={error}
          onChangeText={(input) => setInsert(input)}
          autoCorrect={false}
          placeholder="Adicionar novo repositÃ³rio"
        />
        <Submit onPress={handleInsert}>
          <Icon name="plus" size={20} color={black} />
        </Submit>
      </Form>
      <List
        keyboardShouldPersistsTaps="handled"
        data={repositories}
        onRefresh={loadRepositories}
        refreshing={loading}
        keyExtractor={(item) => String(item.id)}
        renderItem={({ item }) => (
          <Repository data={item} />
        )}
      />

    </Container>
  );

};

export default Repositories;
