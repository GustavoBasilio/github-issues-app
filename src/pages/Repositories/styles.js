import React from 'react';
import styled from 'styled-components';

import {white, grayLight, grayDark, red} from '../../constants/colors';

export const Container = styled.View`
    background: ${grayLight};
    flex: 1;
    padding: 0 15px;
`
export const Form = styled.View`
    display: flex;
    flex-direction: row;
    padding: 15px 0 10px 0;
    margin-bottom: 15px;
    border-style: solid;
    border-bottom-width: 2px;
    border-bottom-color:${grayDark};
`;

export const Input = styled.TextInput.attrs({
    placeholderTextColor: grayDark,
})`
    background-color: ${white};
    padding: 5px 15px;
    flex: 1;
    border-radius: 6px;
    border: 2px solid ${(props) => (props.error ? red : grayDark)};
    font-size: 14px;
`;

export const Submit = styled.TouchableOpacity`
    justify-content: center;
    padding-left: 15px;
`;

export const List = styled.TouchableOpacity.attrs({
    showsVerticalScrollIndicator: false
})`
    padding-top: 10px;
`;