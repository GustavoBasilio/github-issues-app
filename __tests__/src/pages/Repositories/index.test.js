/**
 * @format
 */

import 'react-native';
import React from 'react';
import Repositories from '../../../../src/pages/Repositories/index';

import 'jest-styled-components'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const comp = renderer.create(<Repositories />).toJSON();
  expect(comp).toMatchSnapshot();
});
