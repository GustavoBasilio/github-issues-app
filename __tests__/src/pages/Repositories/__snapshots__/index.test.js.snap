// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#EEEEEE",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#E5E5E5",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "display": "flex",
          "flexDirection": "row",
          "marginBottom": 15,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 15,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      error={false}
      onChangeText={[Function]}
      placeholder="Adicionar novo repositório"
      placeholderTextColor="#E5E5E5"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#E5E5E5",
            "borderRadius": 6,
            "borderStyle": "solid",
            "borderWidth": 2,
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 14,
            "paddingBottom": 5,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 5,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Object {
            "justifyContent": "center",
            "paddingLeft": 15,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </TouchableOpacity>
  </View>
  <TouchableOpacity
    activeOpacity={0.2}
    data={Array []}
    keyExtractor={[Function]}
    keyboardShouldPersistsTaps="handled"
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
    showsVerticalScrollIndicator={false}
    style={
      Array [
        Object {
          "paddingTop": 10,
        },
      ]
    }
  />
</View>
`;
